### ISSUE TITLE
Rename Does not work In Some Cases

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/TypeScript#14278]
From https://github.com/Microsoft/vscode/issues/21319

**TypeScript Version:**  2.2.1 / nightly (2.2.0-dev.201xxxxx)

**Code**
For the JavaScript:

```ts
const foo = {
    set: function (x) {
        this._x = x;
    },
    copy: function (x) {
        this._x = x.prop;
    }
}
```

Trigger a rename on `x` in the `copy` function.

**Expected behavior:**
Rename updates both the parameter name and `x.prop`

**Actual behavior:**
Rename only renames the parameter, not its usage in the body of the function:

```
[Trace - 4:41:49 PM] Sending request: rename (1045). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "/Users/matb/projects/sand/a.js",
    "line": 5,
    "offset": 22,
    "findInStrings": false,
    "findInComments": false
}
[Trace - 4:41:49 PM] Response received: rename (1045). Request took 5 ms. Success: true 
Result: {
    "info": {
        "canRename": true,
        "kind": "parameter",
        "displayName": "x",
        "fullDisplayName": "x",
        "kindModifiers": "",
        "triggerSpan": {
            "start": 86,
            "length": 1
        }
    },
    "locs": [
        {
            "file": "/Users/matb/projects/sand/a.js",
            "locs": [
                {
                    "start": {
                        "line": 5,
                        "offset": 21
                    },
                    "end": {
                        "line": 5,
                        "offset": 22
                    }
                }
            ]
        }
    ]
}
```

Rename works as expected in `set`

**Comment by mhegazy** [comment]
should be fixed by https://github.com/Microsoft/TypeScript/pull/14141 under --strictThisChecks. @sandersn can you please follow up on this?

**Comment by sandersn** [comment]
Some observations:

1. This doesn't repro in typescript. 
2. If you rename `x/**/.prop`, then *only* `this._x` renames (incorrectly) &mdash; neither `function (x/**/)` nor `x/**/.prop` do.
3. If this is a recent regression, it's probably got something to do with the code that looks for assignments to `this` inside functions in Javascript.
4. I don't think it has anything to do with #14141 because that only affects the contextual type of `this`, not the symbol that it uses.

**Comment by sandersn** [comment]
The problem is that `this._x = x.prop` is recognised as a special property assignment in the binder, but in the checker, `getSpecialPropertyAssignmentSymbolFromEntityName` doesn't check which side the PropertyAccessExpression `x.prop` is from. So it treats `this._x` and `x.prop` the same even though only `this._x` should return the special symbol `_x`.

I think the fix is to make `getSpecialPropertyAssignmentSymbolFromEntityName` smarter; the binder behaviour is a little weird, attaching the symbol to the assignment node, but not outright wrong.

**Comment by sandersn** [comment]
Fix is up at #14302

**Comment by v1nce** [comment]
I'm not sure I understand the relationships between typescript and vscode.
Is this supposed to be fixed in vscode 1.10.0-insider / 2017-02-27T14:13:35.653Z ?

**Comment by mhegazy** [comment]
VSCode ships with an older version of TypeScript. Please see [Using Newer TypeScript Versions](https://code.visualstudio.com/docs/languages/typescript#_using-newer-typescript-versions) documentation for more details on updating your VSCode to use a different version of TypeScript.
